### PR TITLE
Add hit-testing to chart

### PR DESCRIPTION
**Pull request by laptou**
- Chart now responds by highlighting a series when you mouse over it

https://drive.google.com/file/d/1-iIHF0s-1zbbeWG0h1yoSHLUFR94PIUm/view?usp=sharing

### CODE REVIEW

File: client/js/chart/index.js [hunk 1]
```diff
@@ -17,15 +17,87 @@ export function createChart(el, rankingHistory, rankingDates) {
...
-  }
+    seriesElements.push(series);
+  });
+
+  let activeSeries = null;
+  let activeX = null;
+  let activeY = null;
+
+  // do some hit-testing
+  svg.addEventListener('mousemove', (ev) => {
```
Comment (seunomonije) [inline client/js/chart/index.js:37]: +1 for the comments here, helps me out alot

File: client/js/chart/index.js [hunk 2]
```diff
@@ -60,6 +132,24 @@ function createSeries(color, history) {
     start = end;
   }
 
+  // create marker for when the user mouses near a point
+  const marker = document.createElementNS(SVG_NS, 'circle');
+  marker.classList.add('series-marker');
+  marker.setAttribute('r', '5px');
```
Comment (gracetian6) [inline client/js/chart/index.js:138]: so there are many (hidden) points on the graph, and circle gets created when user hovers over it?

File: client/js/chart/index.js [hunk 3]
```diff
@@ -17,16 +17,87 @@ export function createChart(el, rankingHistory, rankingDates) {
...
+      // include only series where there is a data point for this x value
+      if (x >= history.length || history[x] === null) return;
+
+      // get distance from mouse y to series y for each series
+      const dist = history[x] - y;
+
+      // filter items where distance is too great; there should only be one
+      // remaining (impossible to be less than 0.4 away from both 1 and 2, for
+      // example)
+      if (Math.abs(dist) > 0.4) return;
```
Comment (gracetian6) [inline client/js/chart/index.js:63]: MATH!!!!

File: client/js/chart/index.js [hunk 4]
```diff
@@ -17,16 +17,87 @@ export function createChart(el, rankingHistory, rankingDates) {
...
+
+      hit = index;
+    });
+
+    if (activeSeries && activeSeries !== hit) {
+      const seriesElement = seriesElements[activeSeries];
+      seriesElement.dispatchEvent(new CustomEvent('series-clear'));
+    }
```
Comment (gracetian6) [inline client/js/chart/index.js:71]: what is `activeSeries`?? Is this condition trying to see whether `activerSeries` exists and has not been hit yet?

File: client/js/chart/index.js [hunk 5]
```diff
@@ -17,16 +17,87 @@ export function createChart(el, rankingHistory, rankingDates) {
...
+    const series = createSeries(svg, color, history);
     seriesContainer.append(series);
-    index++;
-  }
+    seriesElements.push(series);
+  });
+
+  let activeSeries = null;
+  let activeX = null;
+  let activeY = null;
```
Comment (gracetian6) [inline client/js/chart/index.js:34]: can you add a comment for this? Was a bit confused what activeSeries is.

File: client/js/chart/index.js (hunk 2 above)
Comment (laptou) [reply client/js/chart/index.js:138]: The circle gets created at the beginning, but it is added to the DOM when the user hovers, and removed when they move their mouse away

File: client/js/chart/index.js (hunk 4 above)
Comment (laptou) [reply client/js/chart/index.js:71]: It is the series that the user is currently hovering over. How is `currentSeries` a more obvious name than `activeSeries`?

File: client/js/chart/index.js (hunk 5 above)
Comment (laptou) [reply client/js/chart/index.js:34]: changed in #37

**Pull request merged**